lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -934,309 +940,451 @@ bool Inliner::shouldInline(Operation *op) {
+              nlaMap[sym].flattenModule(childModule);
+            else
+              nlaMap[sym].inlineModule(childModule);
+          }
+        }
+
+        // The InstanceOp `instance` might not have a symbol, if it does not
+        // participate in any HierPathOp. But the reTop might add a symbol to
+        // it, if a HierPathOp is added to this Op. If we're about to inline a
+        // module that contains a non-local annotation that starts at that
+        // module, then we need to both update the mutable NLA to indicate that
+        // this has a new top and add an annotation on the instance saying that
+        // this now participates in this new NLA.
+        DenseMap<Attribute, Attribute> symbolRenames;
+        if (!rootMap[childModule.getNameAttr()].empty()) {
+          for (auto sym : rootMap[childModule.getNameAttr()]) {
+            auto &mnla = nlaMap[sym];
+            // Retop to the new parent, which is the topmost module (and not
+            // immediate parent) in case of recursive inlining.
+            sym = mnla.reTop(inlineToParent);
+            StringAttr instSym = getInnerSymName(instance);
+            if (!instSym) {
+              instSym = StringAttr::get(
+                  context, il.mic.modNamespace.newName(instance.getName()));
+              instance.setInnerSymAttr(hw::InnerSymAttr::get(instSym));
+            }
+            instOpHierPaths[InnerRefAttr::get(moduleName, instSym)].push_back(
+                cast<StringAttr>(sym));
+            // TODO: Update any symbol renames which need to be used by the next
+            // call of inlineInto.  This will then check each instance and
+            // rename any symbols appropriately for that instance.
+            symbolRenames.insert({mnla.getNLA().getNameAttr(), sym});
+          }
+        }
+        auto instInnerSym = getInnerSymName(instance);
+        auto parentActivePaths = activeHierpaths;
+        setActiveHierPaths(moduleName, instInnerSym);
+        // This must be done after the reTop, since it might introduce an
+        // innerSym.
+        currentPath.emplace_back(moduleName, instInnerSym);
+
+        InliningLevel childIL(il.mic, childModule);
+        createDebugScope(childIL, instance, il.debugScope);
+
+        // Create the wire mapping for results + ports.
+        auto nestedPrefix = (prefix + instance.getName() + "_").str();
+        mapPortsToWires(nestedPrefix, childIL, mapper, {});
+        mapResultsToWires(mapper, childIL.wires, instance);
+
+        // Inline the module, it can be marked as flatten and inline.
+        if (toBeFlattened) {
+          if (failed(flattenInto(nestedPrefix, childIL, mapper, {})))
+            return failure();
+        } else {
+          if (failed(inlineInto(nestedPrefix, childIL, mapper, symbolRenames)))
+            return failure();
+        }
+        currentPath.pop_back();
+        activeHierpaths = parentActivePaths;
+        return success();
+      });
```

Comment:
I didn't review this part in detail as this all appeared to be code motion. Echoing an earlier comment, this diff may be less heinous without the large anonymous lambda.

Reply:
Thanks!